stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt

```diff
@@ -1026,14 +1053,35 @@ public class ChatClient internal constructor(
         message: Message,
     ): Call<Message> {
         val relevantPlugins = plugins.filterIsInstance<SendMessageListener>()
-
-        return api.sendMessage(channelType, channelId, message)
-            .doOnStart(scope) { relevantPlugins.forEach { it.onMessageSendRequest(channelType, channelId, message) } }
-            .doOnResult(scope) { result ->
-                relevantPlugins.forEach {
-                    it.onMessageSendResult(result, channelType, channelId, message)
+        val relevantInterceptors = interceptors.filterIsInstance<SendMessageInterceptor>()
+        return CoroutineCall(scope) {
+
+            // Message is first prepared i.e. all its attachments are uploaded and message is updated with these attachments.
+            // TODO: An InterceptedCall wrapper can be created to avoid so much code here.
```

Comment:
You can simplify the code like this by using flatMapSuspend instead of checking if result is success:
```
val relevantPlugins = plugins.filterIsInstance<SendMessageListener>()
val relevantInterceptors = interceptors.filterIsInstance<SendMessageInterceptor>()
return CoroutineCall(scope) {
    // Message is first prepared i.e. all its attachments are uploaded and message is updated with these attachments.
    relevantInterceptors.fold(Result.success(message)) { message, interceptor ->
        if (message.isSuccess) {
            interceptor.interceptMessage(channelType, channelId, message.data())
        } else message
    }.flatMapSuspend { newMessage ->
        api.sendMessage(channelType, channelId, newMessage)
            .retry(scope, retryPolicy)
            .doOnResult(scope) { result ->
                relevantPlugins.forEach {
                    it.onMessageSendResult(
                        result,
                        channelType,
                        channelId,
                        newMessage
                    )
                }
            }.await()
    }
}
```

Reply:
Good suggestion.